src/main/java/com/google/sps/scheduler/Scheduler.java

```diff
@@ -0,0 +1,150 @@
+// limitations under the License.
+
+package com.google.sps.scheduler;
+
+import static converter.TimeConverter.epochInMilliseconds;
+import static converter.TimeConverter.epochToDateTime;
+
+import com.google.api.client.util.DateTime;
+import com.google.api.services.calendar.model.Event;
+import com.google.api.services.calendar.model.EventDateTime;
+import com.google.common.collect.ImmutableList;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.Comparator;
+import java.util.concurrent.TimeUnit;
+
+/** Class that allows to schedule some tasks in the free time slot of the calendar. */
+public class Scheduler {
+  public static final int START_HOUR = 9;
+  public static final int START_MINUTE = 0;
+  public static final int END_HOUR = 18;
+  public static final int END_MINUTE = 0;
+  public static final long DEFAULT_DURATION_IN_MILLISECONDS = TimeUnit.MINUTES.toMillis(30);
+  public static final ImmutableList<String> TASKS_SAMPLE = ImmutableList.of(
+      "Review code", "Write the design doc", "Talk to PM", "Investigate report", "Prepare slides");
+
+  /**
+   * Schedules the tasks in the free time slot of the calendar events, which must be of the same day
+   * specified in the parameter.
+   * For each task that can be scheduled, a new calendar event object is created and a list of
+   * all of them is returned.
+   */
+  public static List<Event> schedule(
+      List<Event> calendarEvents, List<String> tasks, String timeZone, LocalDate dayDate) {
+    List<Event> tasksEvents = new ArrayList<>();
+
+    Set<Event> orderedCalendarEvents = new TreeSet<>(
+        Comparator.comparingLong(event -> event.getStart().getDateTime().getValue()));
+    orderedCalendarEvents.addAll(calendarEvents);
+
+    long dayStartEpochMilliseconds =
+        epochInMilliseconds(dayDate, LocalTime.of(START_HOUR, START_MINUTE), timeZone);
+    long dayEndEpochMilliseconds =
+        epochInMilliseconds(dayDate, LocalTime.of(END_HOUR, END_MINUTE), timeZone);
+
+    long lastEnd = dayStartEpochMilliseconds;
+    int indexNextTask = 0;
+
+    for (Event event : orderedCalendarEvents) {
+      long eventEnd = event.getEnd().getDateTime().getValue();
+      if (eventEnd <= dayStartEpochMilliseconds) {
+        continue;
+      }
+
+      long eventStart = event.getStart().getDateTime().getValue();
+      if (eventStart >= dayEndEpochMilliseconds) {
+        break;
+      }
+
+      while (eventStart - lastEnd >= DEFAULT_DURATION_IN_MILLISECONDS &&
+          indexNextTask < tasks.size()) {
+        DateTime startTime = epochToDateTime(lastEnd, timeZone);
+        lastEnd += DEFAULT_DURATION_IN_MILLISECONDS;
+        DateTime endTime = epochToDateTime(lastEnd, timeZone);
+        tasksEvents.add(
+            createEventWithSummary(startTime, endTime, timeZone, tasks.get(indexNextTask)));
+        indexNextTask++;
```

Comment:
you can instead us an [Iterator](https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html) which you get by calling [`tasks.iterator()`](https://docs.oracle.com/javase/8/docs/api/java/util/List.html#iterator--)